libkernelflinger/security_efi.c

```diff
@@ -98,51 +96,92 @@ EFI_STATUS set_device_security_info(__attribute__((unused)) IN void *security_da
 	BOOTLOADER_SEED_PROTOCOL *bls;
 	UINT8 key_count = BLS_MAX_RPMB_KEY;
 	UINT8 rpmb_keys[BLS_MAX_RPMB_KEY][RPMB_KEY_SIZE];
-#ifdef USE_TPM
 	UINT8 rpmb_key[RPMB_KEY_SIZE];
 	UINTN key_size = sizeof(rpmb_key);
-#endif
 	UINT8 i;
 
 	// Set the fixed RPMB key
 	if (is_live_boot()) {
+		// For USB live boot case, always use one fixed RPMB key and seed, and do not store the lock/unlock state.
+		set_secure_storage_type(SECURE_STORAGE_NONE);
+		return set_rpmb_derived_key(fixed_rpmb_keys, RPMB_KEY_SIZE, 1);
+	}
+
+	if (!is_platform_secure_boot_enabled()) {
+		// For secure boot disabled case, always use one fixed RPMB key and seed, and store the lock/unlock state to EFI variable.
+		set_secure_storage_type(SECURE_STORAGE_EFI_VAR);
 		// For USB live boot case, always use one fixed RPMB key.
 		return set_rpmb_derived_key(fixed_rpmb_keys, RPMB_KEY_SIZE, 1);
 	}
 
+	// Try BLS
 	bls = get_bls_proto();
 	if (bls) {
 		ret = uefi_call_wrapper(bls->GetRpmbKey, 2, &key_count, bls_rpmb_key);
-		if (EFI_ERROR(ret)) {
-			efi_perror(ret, L"call GetRpmbKey of bootloader seed protocol fail");
-			ret = set_rpmb_derived_key(fixed_rpmb_keys,
-				sizeof(fixed_rpmb_keys), ARRAY_SIZE(fixed_rpmb_keys));
-			return ret;
+		if (!EFI_ERROR(ret)) {
+			// Success get the RPMB key
+			debug(L"call GetRpmbKey of bootloader seed protocol success");
+			// Check whether the physical RPMB is exist
+			ret = rpmb_init(get_boot_device_handle());
+			if (!EFI_ERROR(ret)) {
+				for(i = 0; i < key_count; i++)
+					memcpy(rpmb_keys[i], bls_rpmb_key[i].rpmb_key, RPMB_KEY_SIZE);
```

Comment:
This is bad to use RPMB_KEY_SIZE.  You should instead use `sizeof(*rpmb_keys)` in order to bind to the actual size and also to not have to update this code if the `RPMB_KEY_SIZE` constant name changes.